runtime/src/multi_sig.rs

```diff
@@ -0,0 +1,413 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::identity;
+use codec::{Decode, Encode};
+use primitives::{AuthorizationData, AuthorizationError, Key, Signer};
+use rstd::{convert::TryFrom, prelude::*};
+use sr_primitives::{
+    traits::{Dispatchable, Hash},
+    weights::{GetDispatchInfo, Weight},
+    DispatchError,
+};
+use srml_support::{decl_event, decl_module, decl_storage, dispatch::Result, ensure, StorageValue};
+use system::ensure_signed;
+
+pub trait Trait: system::Trait + identity::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Multisig {
+        /// Nonce to ensure unique Multisig addresses are generated. starts from 1.
+        pub MultiSigNonce get(ms_nonce) build(|_| 1u64): u64;
+
+        /// Signers of a multisig. (mulisig, signer) => true/false
+        pub MultiSigSigners get(ms_signers): map (T::AccountId, Signer) => bool;
+        /// Confirmations required before processing a multisig tx
+        pub MultiSigSignsRequired get(ms_signs_required): map T::AccountId => u64;
+        /// Number of transactions proposed in a multisig. Used as tx id. starts from 0
+        pub MultiSigTxDone get(ms_tx_done): map T::AccountId => u64;
+
+        /// Proposals presented for voting to a multisig (multisig, proposal id) => Option<proposal>.
+        /// It is deleted after proposal is processed
+        pub Proposals get(proposals): map (T::AccountId, u64) => Option<T::Proposal>;
+
+        /// Number of votes in favor of a tx. Mapping from (multisig, tx id) => no. of approvals.
+        pub TxApprovals get(tx_approvals): map (T::AccountId, u64) => u64;
+        /// Individual multisig signer votes. (multi sig, signer, )
+        pub Votes get(votes): map (T::AccountId, Signer, u64) => bool;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event() = default;
+
+        /// Creates a multisig
+        ///
+        /// # Arguments
+        /// * `signers` - Signers of the multisig (They need to accept authorization before they are actually added).
+        /// * `sigs_required` - Number of sigs required to process a multi-sig tx.
+        pub fn create_multi_sig(origin, signers: Vec<Signer>, sigs_required: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+            ensure!(signers.len() > 0, "No signers provided");
+            ensure!(u64::try_from(signers.len()).unwrap_or_default() >= sigs_required && sigs_required > 0,
+                "Sigs required out of bounds"
+            );
+            let nonce: u64 = Self::ms_nonce();
+            let new_nonce: u64 = nonce + 1u64;
+            <MultiSigNonce>::put(new_nonce);
+
+            let h: T::Hash = T::Hashing::hash(&(b"MULTI_SIG", new_nonce, sender.clone()).encode());
```

Comment:
Why does the multisig wallet ID need to be connected to the sender via the hash? Cannot `wallet_id` simply be a fresh `T::AccountID`?

Reply:
There's no direct way to generate "fresh" `T::AccountID`. There is no central module or anything that stores `T::AccountID`. `T::AccountID` is just a hash and must be unique for all accounts. It's usually the public key of the transactor but since multi sigs do not have a public key, this is a hack to assign them a unique accountid. `new_nonce` guarantees uniqueness and `sender.clone()` prevents someone from pre-computing all multi-sig addresses/accountids.

Reply:
You are right, there is currently no way to generate a `T::AccountId` from initial data without using the codec. Nevertheless, instead of inlining its generation two times, would it be possible to have a special `multi_sig::Trait::AccountId` with an instance of `From<(u64, Signer)>` or perhaps just a `new` method? If that is done and compartmentalised, additional issues like using something apart from `Codec` for account ID generation could be phrased out later.

Reply:
I don't want to define a custom type/trait here because I want the multi-sig ids to be interchangeable with the normal AccountIds. I can perhaps define a new/from function over the generic AccountIds to allow this conversion directly but since this code is not going to be used at any other place, I don't think it's worth doing.

Reply:
Would something like this work?
```
struct AccountId<T: Trait>(T::AccountId);

impl<T> TryFrom<(AccountId<T>, u64)> for AccountId<T>
where
    T: Trait,
{
    type Error = &'static str;

    fn try_from(
        (AccountId(account_id), nonce): (AccountId<T>, u64),
    ) -> rstd::result::Result<Self, Self::Error> {
        let h = T::Hashing::hash(&(b"MULTI_SIG", nonce, account_id).encode());
        T::AccountId::decode(&mut &h.encode().as_slice()[..])
            .map_err(|_| "Error in decoding multisig address")
            .map(AccountId)
    }
}
```
Perhaps even with getting and updating the nonce within?

Reply:
I don't see the need for going through all this trouble. Using `try_from` for this use-case doesn't sound right either because we are not converting from (AccountId<T>, u64) -> AccountId<T>. We are modifying the data to generate a fresh AccountId and hence a `new` function will suit here better. That being said, I just don't see the point of making things more complex than they already are. There's already a function where you pass the sender and it returns the (next) multi_sig address. I don't think we'll be needing this/similar code anywhere else.

Reply:
I just made an example. How to factor out those two identical encode-decode constructions in `create_multisig` and `gen_next_multisig_address` is your choice ;-)

BTW the struct was only a way to implement the `TryFrom` trait because as I said there is no straightforward way to add traits to `system::Trait::AccountId`.

Reply:
https://github.com/PolymathNetwork/Polymesh/pull/142/commits/3b41a1054f6fbd10b1b27316c20c7826ad88ef17 Please check if this is acceptable.

Reply:
It's very good, thanks! As a minor comment, I'm not sure why it's public because it's only used within that module.

Reply:
Is there any drawback of making it public? I tend to make all utility functions that *may* be used by other modules in the future, public.

Reply:
I can think of only one difference: unused public functions don't cause a warning and I think are kept in the target library.

Reply:
I think that's fine for our use case. That doesn't have any runtime performance penalty and allows us to use this utility function in external modules/crates. We are planning on splitting all of the modules in their separate crates anyway.